spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteManifestsAction.java

```diff
@@ -466,6 +474,309 @@ public void testRewriteLargeManifestsPartitionedTable() throws IOException {
+
+    // Each file should contain less than the full c3 partition span
+    c3Boundaries.forEach(
+        boundary -> {
+          assertThat(boundary.second() - boundary.first())
+              .as("Manifest should contain less than the full range of c3 bucket partitions")
+              .isLessThanOrEqualTo(c3PartitionMax - c3PartitionMin);
+        });
+
+    // c3's Bucket(10) partition means our true lower bound = 0 and true upper bound is 9. The first
+    // manifest should
+    // include the lower bound of 0, and the last should have the upper bound of 9
+    assertThat(lowers.get(0))
+        .withFailMessage("Lower bound of first manifest partition should be 0")
+        .isEqualTo(c3PartitionMin);
+    assertThat(uppers.get(uppers.size() - 1))
+        .withFailMessage("Lower bound of first manifest partition should be 0")
+        .isEqualTo(c3PartitionMax);
+  }
+
+  @TestTemplate
+  public void testRewriteManifestsPartitionedTableWithCustomSortFunction() throws IOException {
+    Random random = new Random();
+
+    PartitionSpec spec =
+        PartitionSpec.builderFor(SCHEMA).identity("c1").truncate("c2", 3).bucket("c3", 10).build();
+    Table table = TABLES.create(SCHEMA, spec, tableLocation);
+
+    List<DataFile> dataFiles = Lists.newArrayList();
+    for (int fileOrdinal = 0; fileOrdinal < 1000; fileOrdinal++) {
+      dataFiles.add(
+          newDataFile(
+              table,
+              TestHelpers.Row.of(
+                  new Object[] {
+                    fileOrdinal, String.valueOf(random.nextInt() * 100), random.nextInt(10)
+                  })));
+    }
+    ManifestFile appendManifest = writeManifest(table, dataFiles);
+    table.newFastAppend().appendManifest(appendManifest).commit();
+
+    List<ManifestFile> manifests = table.currentSnapshot().allManifests(table.io());
+    assertThat(manifests).as("Should have 1 manifests before rewrite").hasSize(1);
+
+    // Capture the c3 partition's lower and upper bounds - used for later test assertions
+    Integer c3PartitionMin =
+        Conversions.fromByteBuffer(
+            Types.IntegerType.get(), manifests.get(0).partitions().get(2).lowerBound());
+    Integer c3PartitionMax =
+        Conversions.fromByteBuffer(
+            Types.IntegerType.get(), manifests.get(0).partitions().get(2).upperBound());
+
+    // Set the target manifest size to a small value to force splitting records into multiple files
+    table
+        .updateProperties()
+        .set(
+            TableProperties.MANIFEST_TARGET_SIZE_BYTES,
+            String.valueOf(manifests.get(0).length() / 3))
+        .commit();
+
+    SparkActions actions = SparkActions.get();
+
+    // This is the main point of this test!
+    // Let's say I want to sort manifests into buckets 0-4 and 5 - 9
+    // I know my Table and partition Spec, so can do that programmatically
+    Function<DataFile, String> test =
+        (Function<DataFile, String> & Serializable)
```

Comment:
I could not figure out a way to pass a `Function` around for use in UDFs without declaring it like this at the outset. Looking for ideas here, this part of the API is a bit clunky.